src/librustdoc/clean/types.rs

```diff
@@ -48,7 +49,82 @@ use self::ItemKind::*;
 use self::SelfTy::*;
 use self::Type::*;
 
-thread_local!(crate static MAX_DEF_IDX: RefCell<FxHashMap<CrateNum, DefIndex>> = Default::default());
+crate type DefIdSet = FxHashSet<DefId>;
+crate type RealDefId = rustc_hir::def_id::DefId;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Copy)]
+crate enum DefId {
+    Real(RealDefId),
+    Fake(usize),
+}
+
+impl DefId {
+    crate fn new_fake() -> Self {
```

Comment:
nit: this will cause a bunch of issues with `cfg(parallel_compiler)` set - can you change it to unconditionally panic instead so we know to change it?
```suggestion
    #[cfg(parallel_compiler)]
    crate fn new_fake() -> Self { unimplemented!() }
    #[cfg(not(parallel_compiler))]
    crate fn new_fake() -> Self {
```

Reply:
What does `parallel_compiler` do?

Reply:
It makes the compiler multithreaded. I *think* right now that only affects queries, but I would rather panic than silently have the wrong behavior.

Reply:
Ah yes and Thread local would destroy the uniqueness. Couldn't we just use a static?

Reply:
Sure, that works too. But parallel-compiler isn't really developed right now so I'm fine with just not supporting it.